File: 2020-hunan/rectangle-union-exepctation/_solution.cc
#include <cstdint>
#include <cstdio>
#include <algorithm>

static const int MOD = 1e9 + 7;

void add(int &x, int a) {
  x += a;
  if (x >= MOD) {
    x -= MOD;
  }
}

int sum(int a, int b)
{
  return (uint64_t)(a + b) * (b - a + 1) / 2 % MOD;
}

int solve(int a, int x1, int x2)
{
  int result = 0;
  // if i < x1, i + (x1 - x2)
  // if x1 <= i < x2, x2
  // if x2 <= i, i
  if (x2 <= a) {
    add(result, sum(x2, a));
  }
}

int main() {
  int a, b;
  while (scanf("%d%d", &a, &b) == 2) {
    int x1, x2, y1, y2;
    scanf("%d%d%d%d", &x1, &x2, &y1, &y2);
    printf("%d\n", (int)((uint64_t)solve(a, x1, x2) * solve(b, y1, y2) % MOD));
  }
}

File: 2020-hunan/rectangle-union-exepctation/solution.py
# from __future__ import print_function

MOD = 10 ** 9 + 7

def asum(a, b):
  return (a + b) * (b - a + 1) // 2

def solve(a, x1, x2):
  if a <= x1:
    return 0
  result = 0
  if a >= x2:
    result += (x2 - x1) * (a - x2 + 1)
  result += asum(1, min(a, x2 - 1) - x1)
  return result

while True:
  try:
    a, b = list(map(int, input().split()))
    x1, x2, y1, y2 = list(map(int, input().split()))
    print((a * b * (x2 - x1) * (y2 - y1) + asum(1, a) * asum(1, b) - solve(a, x1, x2) * solve(b, y1, y2)) % MOD)
  except EOFError:
    break
